packages/google_maps_flutter/lib/src/controller.dart

```diff
@@ -144,95 +110,76 @@ class GoogleMapController extends ChangeNotifier {
-  /// platform side.
-  ///
-  /// The returned [Future] completes with the added marker once listeners have
-  /// been notified.
-  Future<Marker> addMarker(MarkerOptions options) async {
-    final MarkerOptions effectiveOptions =
-        MarkerOptions.defaultOptions.copyWith(options);
-    // TODO(amirh): remove this on when the invokeMethod update makes it to stable Flutter.
-    // https://github.com/flutter/flutter/issues/26431
-    // ignore: strong_mode_implicit_dynamic_method
-    final String markerId = await _channel.invokeMethod(
-      'marker#add',
-      <String, dynamic>{
-        'options': effectiveOptions._toJson(),
-      },
-    );
-    final Marker marker = Marker(markerId, effectiveOptions);
-    _markers[markerId] = marker;
+/// Manages lifecycle of [MarkerController] for all [Marker]s.
+///
+/// Change listeners are notified upon changes to any of the markers.
+class MarkerControllers extends ChangeNotifier {
+  MarkerControllers();
+
+  final Map<MarkerId, MarkerController> _markerControllers =
+      <MarkerId, MarkerController>{};
+
+  void update(MarkerUpdates markerUpdates) {
+    markerUpdates.markerUpdates.forEach((MarkerUpdate markerUpdate) {
+      final MarkerId markerId = markerUpdate.markerId;
+      switch (markerUpdate.updateEventType) {
+        case MarkerUpdateEventType.update:
+          _markerControllers[markerId].setMarker(markerUpdate.newMarker);
+          break;
+        case MarkerUpdateEventType.add:
+          final MarkerController controller =
+              MarkerController.init(markerUpdate.changes);
+          _markerControllers[markerId] = controller;
+          break;
+        case MarkerUpdateEventType.remove:
+          // TODO (kaushik) any other channel cleanup?
+          _markerControllers.remove(markerId);
+          break;
+        default:
+          throw Exception("Unknown markerUpdate type.");
+      }
+    });
     notifyListeners();
-    return marker;
   }
+}
 
-  /// Updates the specified [marker] with the given [changes]. The marker must
-  /// be a current member of the [markers] set.
-  ///
-  /// Change listeners are notified once the marker has been updated on the
-  /// platform side.
-  ///
-  /// The returned [Future] completes once listeners have been notified.
-  Future<void> updateMarker(Marker marker, MarkerOptions changes) async {
-    assert(marker != null);
-    assert(_markers[marker._id] == marker);
-    assert(changes != null);
-    // TODO(amirh): remove this on when the invokeMethod update makes it to stable Flutter.
-    // https://github.com/flutter/flutter/issues/26431
-    // ignore: strong_mode_implicit_dynamic_method
-    await _channel.invokeMethod('marker#update', <String, dynamic>{
-      'marker': marker._id,
-      'options': changes._toJson(),
-    });
-    marker._options = marker._options.copyWith(changes);
-    notifyListeners();
+/// Handles callbacks for events on [Marker] and [InfoWindow].
+class MarkerController {
```

Comment:
The name MarkerController hints to me that it controls everything related to this marker (e.g icon, position, etc...) this is more of a MarkerEventDelegate